netket/experimental/operator/_fermions_2nd.py

```diff
@@ -0,0 +1,695 @@
+                mels[n_c] = constant
+                n_c += 1
+
+            new_term = True
+            # loop over all terms and sum where necessary
+            for orb_idx, dagger, weight, term_end in zip(
+                orb_idxs, daggers, weights, term_ends
+            ):
+                if new_term:
+                    new_term = False
+                    xt = np.copy(xb)
+                    mel = weight
+                    has_xp = True
+
+                if has_xp:  # previous term items made a zero, so skip
+
+                    empty_site = is_empty(xt[orb_idx])
+
+                    if dagger:
+                        if not empty_site:
+                            has_xp = False
+                        else:
+                            mel *= (-1) ** np.sum(xt[:orb_idx])  # jordan wigner sign
+                            xt[orb_idx] = flip(xt[orb_idx])
+                    else:
+                        if empty_site:
+                            has_xp = False
+                        else:
+                            mel *= (-1) ** np.sum(xt[:orb_idx])  # jordan wigner sign
+                            xt[orb_idx] = flip(xt[orb_idx])
+
+                # if this is the end of the term, we collect things
+                if term_end:
+
+                    if has_xp:
+                        x_prime[n_c, :] = xt
+                        mels[n_c] = mel
+                        n_c += 1
+
+                    new_term = True
+
+            if pad:
+                n_c = (b + 1) * max_conn
+
+            sections[b] = n_c
+
+        if pad:
+            return x_prime, mels
+        else:
+            return x_prime[:n_c], mels[:n_c]
+
+    def _op__matmul__(self, other):
+        if not isinstance(other, FermionOperator2nd):
+            return NotImplementedError
+        if not self.hilbert == other.hilbert:
+            raise ValueError(
+                f"Can only multiply identical hilbert spaces (got A@B, A={self.hilbert}, B={other.hilbert})"
+            )
+        terms = []
+        weights = []
+        for t, w in self._operators.items():
+            for to, wo in other._operators.items():
+                terms.append(tuple(t) + tuple(to))
+                weights.append(w * wo)
+        if other._constant != 0.0:
+            for t, w in self._operators.items():
+                terms.append(tuple(t))
+                weights.append(w * other._constant)
+        if self._constant != 0:
+            for t, w in other._operators.items():
+                terms.append(tuple(t))
+                weights.append(w * self._constant)
+        constant = self._constant * other._constant
+        dtype = np.promote_types(self.dtype, other.dtype)
+        return FermionOperator2nd(
+            self.hilbert, terms, weights=weights, constant=constant, dtype=dtype
+        )
+
+    def __radd__(self, other):
+        return self + other
+
+    def __add__(self, other):
+        op = self.copy()
```

Comment:
you should promote types.

```suggestion
        op = self.copy(dtype=np.promote_types(self.dtype, _dtype(other)))
```